File: apps/project-generation/pkg/generator/engine.go
package generator

import (
	"bytes"
	"fmt"
	"io"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
	"text/template"
	"time"
)

// Engine manages project generation from DDD templates.
type Engine struct {
	baseTemplateDir string
}

// NewEngine creates a new generator engine instance.
func NewEngine(baseTemplateDir ...string) *Engine {
	dir := ""
	if len(baseTemplateDir) > 0 {
		dir = baseTemplateDir[0]
	}
	return &Engine{
		baseTemplateDir: dir,
	}
}

// FindTemplateDir locates the template directory by subfolder name.
func (e *Engine) FindTemplateDir(subDir, customDir string) (string, error) {
	if customDir != "" {
		target := filepath.Join(customDir, subDir)
		if fi, err := os.Stat(target); err == nil && fi.IsDir() {
			return target, nil
		}
		if fi, err := os.Stat(customDir); err == nil && fi.IsDir() {
			return customDir, nil
		}
	}

	if e.baseTemplateDir != "" {
		target := filepath.Join(e.baseTemplateDir, subDir)
		if fi, err := os.Stat(target); err == nil && fi.IsDir() {
			return target, nil
		}
	}

	candidateBases := []string{
		"templates",
		"../templates",
		"../../templates",
		"apps/project-generation/templates",
		"../apps/project-generation/templates",
	}

	for _, base := range candidateBases {
		target := filepath.Join(base, subDir)
		if fi, err := os.Stat(target); err == nil && fi.IsDir() {
			return target, nil
		}
	}

	return "", fmt.Errorf("template directory '%s' not found", subDir)
}

// ResolveTemplateDir locates the template directory across common search paths.
func (e *Engine) ResolveTemplateDir(lang SupportedLanguage, customDir string) (string, error) {
	var subDir string
	switch lang {
	case LanguageGo:
		subDir = "microservice-go-ddd"
	case LanguagePython:
		subDir = "microservice-python-ddd"
	case LanguageK8s:
		subDir = "k8s-manifests"
	default:
		return "", fmt.Errorf("unsupported language: %s", lang)
	}

	return e.FindTemplateDir(subDir, customDir)
}

// Generate creates a new project directory and renders all files from the template.
func (e *Engine) Generate(cfg ProjectConfig) (*GenerationResult, error) {
	startTime := time.Now()

	if cfg.ProjectName == "" {
		cfg.ProjectName = "my-service"
	}
	if cfg.ModuleName == "" {
		cfg.ModuleName = cfg.ProjectName
	}
	if cfg.ServiceName == "" {
		cfg.ServiceName = cfg.ProjectName
	}
	if cfg.Port == "" {
		cfg.Port = "8080"
	}
	if cfg.GRPCPort == "" {
		cfg.GRPCPort = "9090"
	}
	if cfg.Database == "" {
		cfg.Database = "postgres"
	}
	if cfg.Description == "" {
		cfg.Description = fmt.Sprintf("%s microservice with DDD architecture", cfg.ProjectName)
	}
	if cfg.Author == "" {
		cfg.Author = "Elmos Platform Team"
	}
	if cfg.Version == "" {
		cfg.Version = "0.1.0"
	}
	if cfg.Environment == "" {
		cfg.Environment = "development"
	}
	if cfg.Namespace == "" {
		cfg.Namespace = "default"
	}
	if cfg.ImageRepository == "" {
		cfg.ImageRepository = cfg.ProjectName
	}
	if cfg.ImageTag == "" {
		cfg.ImageTag = "latest"
	}
	if cfg.Replicas <= 0 {
		cfg.Replicas = 2
	}
	if cfg.OutputDir == "" {
		cfg.OutputDir = filepath.Join(".", cfg.ProjectName)
	}

	tmplDir, err := e.ResolveTemplateDir(cfg.Language, cfg.TemplateDir)
	if err != nil {
		return nil, err
	}

	withTelemetry := true
	if cfg.WithTelemetry != nil {
		withTelemetry = *cfg.WithTelemetry
	}
	withResilience := true
	if cfg.WithResilience != nil {
		withResilience = *cfg.WithResilience
	}

	vars := TemplateVars{
		ProjectName:     cfg.ProjectName,
		ModuleName:      cfg.ModuleName,
		ServiceName:     cfg.ServiceName,
		Port:            cfg.Port,
		GRPCPort:        cfg.GRPCPort,
		Database:        cfg.Database,
		Description:     cfg.Description,
		Author:          cfg.Author,
		Version:         cfg.Version,
		Environment:     cfg.Environment,
		Replicas:        cfg.Replicas,
		Namespace:       cfg.Namespace,
		ImageRepository: cfg.ImageRepository,
		ImageTag:        cfg.ImageTag,
		WithTelemetry:   withTelemetry,
		WithResilience:  withResilience,
	}

	var generatedFiles []string

	// Walk the template directory
	err = filepath.WalkDir(tmplDir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}

		// Calculate relative path inside template directory
		relPath, err := filepath.Rel(tmplDir, path)
		if err != nil {
			return err
		}

		if relPath == "." {
			return nil
		}

		// Skip optional feature files if disabled
		cleanRel := filepath.ToSlash(relPath)
		if !withTelemetry && strings.Contains(cleanRel, "telemetry") {
			if d.IsDir() {
				return filepath.SkipDir
			}
			return nil
		}
		if !withResilience && (strings.Contains(cleanRel, "resilience") || strings.Contains(cleanRel, "circuit_breaker")) {
			if d.IsDir() {
				return filepath.SkipDir
			}
			return nil
		}

		// Target output path
		outRelPath := relPath
		isTemplate := strings.HasSuffix(relPath, ".tmpl")
		if isTemplate {
			outRelPath = strings.TrimSuffix(relPath, ".tmpl")
		}

		targetPath := filepath.Join(cfg.OutputDir, outRelPath)

		if d.IsDir() {
			return os.MkdirAll(targetPath, 0755)
		}

		// Ensure parent directory exists
		if err := os.MkdirAll(filepath.Dir(targetPath), 0755); err != nil {
			return fmt.Errorf("failed to create directory %s: %w", filepath.Dir(targetPath), err)
		}

		// Render or copy file
		if isTemplate {
			if err := renderTemplateFile(path, targetPath, vars); err != nil {
				return fmt.Errorf("error rendering template %s: %w", path, err)
			}
		} else {
			if err := copyFile(path, targetPath); err != nil {
				return fmt.Errorf("error copying file %s: %w", path, err)
			}
		}

		generatedFiles = append(generatedFiles, targetPath)
		return nil
	})

	if err != nil {
		return &GenerationResult{
			Success:        false,
			OutputDir:      cfg.OutputDir,
			FilesGenerated: generatedFiles,
			Duration:       time.Since(startTime),
			Error:          err.Error(),
		}, err
	}

	return &GenerationResult{
		Success:        true,
		OutputDir:      cfg.OutputDir,
		FilesGenerated: generatedFiles,
		Duration:       time.Since(startTime),
	}, nil
}

func renderTemplateFile(srcPath, dstPath string, vars TemplateVars) error {
	content, err := os.ReadFile(srcPath)
	if err != nil {
		return err
	}

	tmplFuncs := template.FuncMap{
		"lower": strings.ToLower,
		"upper": strings.ToUpper,
		"title": strings.Title,
		"snake": func(s string) string {
			return strings.ToLower(strings.ReplaceAll(s, "-", "_"))
		},
		"kebab": func(s string) string {
			return strings.ToLower(strings.ReplaceAll(s, "_", "-"))
		},
		"default": func(defVal, val any) any {
			if val == nil || val == "" {
				return defVal
			}
			return val
		},
		"quote": func(val any) string {
			return fmt.Sprintf("%q", fmt.Sprintf("%v", val))
		},
		"replace": func(old, new, src string) string {
			return strings.ReplaceAll(src, old, new)
		},
		"trunc": func(length int, s string) string {
			if len(s) <= length {
				return s
			}
			return s[:length]
		},
		"trimSuffix": func(suffix, s string) string {
			return strings.TrimSuffix(s, suffix)
		},
		"indent": func(spaces int, s string) string {
			pad := strings.Repeat(" ", spaces)
			lines := strings.Split(s, "\n")
			for i, line := range lines {
				if line != "" {
					lines[i] = pad + line
				}
			}
			return strings.Join(lines, "\n")
		},
		"nindent": func(spaces int, s string) string {
			return "\n" + strings.Repeat(" ", spaces) + s
		},
	}

	tmpl, err := template.New(filepath.Base(srcPath)).
		Option("missingkey=zero").
		Funcs(tmplFuncs).
		Parse(string(content))
	if err != nil {
		return fmt.Errorf("parse error: %w", err)
	}

	var buf bytes.Buffer
	dataMap := vars.ToMap()
	if err := tmpl.Execute(&buf, dataMap); err != nil {
		return fmt.Errorf("execute error: %w", err)
	}

	return os.WriteFile(dstPath, buf.Bytes(), 0644)
}

func copyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()

	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer out.Close()

	_, err = io.Copy(out, in)
	return err
}

// RenderStringPlaceholders replaces simple string placeholders like {{.Name}}.
func renderStringPlaceholders(s string, data map[string]interface{}) string {
	res := s
	for k, v := range data {
		res = strings.ReplaceAll(res, fmt.Sprintf("{{.%s}}", k), fmt.Sprintf("%v", v))
	}
	return res
}

// RenderTemplateContent renders arbitrary template string content using map data.
func renderTemplateContent(tmplContent string, data map[string]interface{}) (string, error) {
	tmpl, err := template.New("tmpl").Option("missingkey=zero").Parse(tmplContent)
	if err != nil {
		return "", err
	}
	var buf bytes.Buffer
	if err := tmpl.Execute(&buf, data); err != nil {
		return "", err
	}
	return buf.String(), nil
}

File: apps/project-generation/project_generation/engine.py
"""Project generation engine implementing template rendering and scaffolding."""

import os
import re
import shutil
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional


@dataclass
class ProjectConfig:
    """Configuration options for scaffolding a microservice project."""
    language: str  # "go", "python", "k8s"
    project_name: str = "my-service"
    module_name: str = ""
    service_name: str = ""
    port: str = "8080"
    grpc_port: str = "9090"
    database: str = "postgres"
    description: str = ""
    author: str = "Elmos Platform Team"
    version: str = "0.1.0"
    environment: str = "development"
    namespace: str = "default"
    image_repository: str = ""
    image_tag: str = "latest"
    replicas: int = 2
    output_dir: str = ""
    template_dir: str = ""
    with_telemetry: bool = True
    with_resilience: bool = True

    def __post_init__(self) -> None:
        if not self.module_name:
            self.module_name = self.project_name
        if not self.service_name:
            self.service_name = self.project_name
        if not self.image_repository:
            self.image_repository = self.project_name
        if not self.description:
            self.description = f"{self.project_name} microservice with DDD architecture"
        if not self.output_dir:
            self.output_dir = os.path.join(".", self.project_name)


@dataclass
class GenerationResult:
    """Outcome report of project generation."""
    success: bool
    output_dir: str
    files_generated: List[str] = field(default_factory=list)
    duration_seconds: float = 0.0
    error: Optional[str] = None

    @property
    def output_directory(self) -> str:
        return self.output_dir


class ProjectGenerator:
    """Orchestrates reading templates and rendering output directory structures."""

    def __init__(self, base_template_dir: str = "") -> None:
        self.base_template_dir = base_template_dir

    def resolve_template_dir(self, language: str, custom_dir: str = "") -> Path:
        """Locates template folder across common repository paths."""
        mapping = {
            "go": "microservice-go-ddd",
            "python": "microservice-python-ddd",
            "k8s": "k8s-manifests",
        }
        sub_dir = mapping.get(language.lower())
        if not sub_dir:
            raise ValueError(f"Unsupported language target: '{language}'")

        if custom_dir:
            p = Path(custom_dir)
            if (p / sub_dir).is_dir():
                return p / sub_dir
            if p.is_dir():
                return p

        if self.base_template_dir:
            p = Path(self.base_template_dir)
            if (p / sub_dir).is_dir():
                return p / sub_dir

        candidate_bases = [
            Path("templates"),
            Path("../templates"),
            Path("../../templates"),
            Path("apps/project-generation/templates"),
            Path("../apps/project-generation/templates"),
        ]

        for base in candidate_bases:
            cand = base / sub_dir
            if cand.is_dir():
                return cand.resolve()

        raise FileNotFoundError(f"Template directory for '{language}' ('{sub_dir}') not found.")

    def generate(self, config: ProjectConfig) -> GenerationResult:
        """Scaffold the project into the designated output directory."""
        start_time = time.time()
        generated_files: List[str] = []

        try:
            tmpl_dir = self.resolve_template_dir(config.language, config.template_dir)
            out_root = Path(config.output_dir).resolve()
            out_root.mkdir(parents=True, exist_ok=True)

            variables: Dict[str, Any] = {
                "ProjectName": config.project_name,
                "ModuleName": config.module_name,
                "ServiceName": config.service_name,
                "Port": config.port,
                "GRPCPort": config.grpc_port,
                "Database": config.database,
                "Description": config.description,
                "Author": config.author,
                "Version": config.version,
                "Environment": config.environment,
                "Namespace": config.namespace,
                "ImageRepository": config.image_repository,
                "ImageTag": config.image_tag,
                "Replicas": config.replicas,
                "WithTelemetry": config.with_telemetry,
                "WithResilience": config.with_resilience,
            }

            for root, dirs, files in os.walk(tmpl_dir):
                rel_dir = Path(root).relative_to(tmpl_dir)
                rel_dir_str = str(rel_dir).replace("\\\\", "/")

                # Skip directories if disabled
                if not config.with_telemetry and "telemetry" in rel_dir_str:
                    continue
                if not config.with_resilience and ("resilience" in rel_dir_str or "circuit_breaker" in rel_dir_str):
                    continue

                target_dir = out_root / rel_dir
                target_dir.mkdir(parents=True, exist_ok=True)

                for file in files:
                    if not config.with_telemetry and "telemetry" in file:
                        continue
                    if not config.with_resilience and ("resilience" in file or "circuit_breaker" in file):
                        continue

                    src_file = Path(root) / file
                    is_tmpl = file.endswith(".tmpl")
                    out_filename = file[:-5] if is_tmpl else file
                    dst_file = target_dir / out_filename

                    if is_tmpl:
                        content = src_file.read_text(encoding="utf-8")
                        rendered = self._render_template_text(content, variables)
                        dst_file.write_text(rendered, encoding="utf-8")
                    else:
                        shutil.copy2(src_file, dst_file)

                    generated_files.append(str(dst_file))

            return GenerationResult(
                success=True,
                output_dir=str(out_root),
                files_generated=generated_files,
                duration_seconds=time.time() - start_time,
            )
        except Exception as e:
            return GenerationResult(
                success=False,
                output_dir=config.output_dir,
                files_generated=generated_files,
                duration_seconds=time.time() - start_time,
                error=str(e),
            )

    @staticmethod
    def _render_template_text(text: str, variables: Dict[str, Any]) -> str:
        """Render Go-style and Jinja-style variables in template text."""
        result = text
        for key, val in variables.items():
            # Go template pattern: {{.Key}}
            result = re.sub(rf"\{{\{{\s*\.{key}\s*\}}\}}", str(val), result)
            # Jinja template pattern: {{ Key }}
            result = re.sub(rf"\{{\{{\s*{key}\s*\}}\}}", str(val), result)
            # Default pipe fallback pattern: {{.Key | default "..."}}
            result = re.sub(rf"\{{\{{\s*\.{key}\s*\|\s*default\s+[^}}]+\}}\}}", str(val), result)

        # Handle simple helper functions if present
        def lower_repl(m: re.Match) -> str:
            v = m.group(1).lstrip(".")
            return str(variables.get(v, "")).lower()

        def upper_repl(m: re.Match) -> str:
            v = m.group(1).lstrip(".")
            return str(variables.get(v, "")).upper()

        result = re.sub(r"\{\{\s*lower\s+(\.?[a-zA-Z0-9_]+)\s*\}\}", lower_repl, result)
        result = re.sub(r"\{\{\s*upper\s+(\.?[a-zA-Z0-9_]+)\s*\}\}", upper_repl, result)

        return result
